usr/src/uts/common/fs/zfs/zfs_dir.c

```diff
@@ -609,16 +609,23 @@ zfs_rmnode(znode_t *zp)
 	}
 
 	/*
-	 * Free up all the data in the file.
+	 * Free up all the data in the file.  We don't do this for directories
+	 * because we need truncate and remove to be in the same tx, like in
+	 * zfs_znode_delete(). Otherwise, if we crash here we'll end up with
+	 * an inconsistent truncated zap object in the delete queue.  Note a
```

Comment:
You may need to ask the ZoL committer @tuxoko directly.  I take it, though, you're suggesting something like:

`if (ZTOV(zp)->v_type == VREG || NOT_XATTR_DIR(zp))`

for some check NOT_XATTR_DIR() I need to UTSL and figure out... right?